src/deploy/index.js

```diff
@@ -53,36 +53,54 @@ var deploy = function (targetNames, options, customContext = {}) {
   var postdeploys = [];
   var startTime = Date.now();
 
-  for (var i = 0; i < targetNames.length; i++) {
-    var targetName = targetNames[i];
-    var target = TARGETS[targetName];
-
-    if (!target) {
-      return Promise.reject(
-        new FirebaseError(clc.bold(targetName) + " is not a valid deploy target", { exit: 1 })
-      );
+  return new Promise((resolve) => {
+    if (targetNames.includes("hosting") && options.config.get("hosting.source")) {
+      resolve(require("firebase-frameworks").prepare(targetNames, context, options));
+    } else {
+      resolve();
     }
+  })
+    .then(function () {
+      // Ensure that functions predeploy & deploy gets called before hosting that way
```

Comment:
This seems like a bit of a roundabout way to do things. AFAICT, the order for these is defined here, https://github.com/firebase/firebase-tools/blob/master/src/commands/deploy.js#L15 and it looks like functions is already before hosting. 

Is there some case that I'm missing where we cant just rely on that ordering?

Reply:
Ah I see, it's firebase-frameworks that's just pushing onto the targets array. I'll fix that and drop the ordering force.